sys/storage/ftl/ftl.c

```diff
@@ -0,0 +1,488 @@
+            return -EBADF;
+        } else if(result == Hamming_ERROR_SINGLEBIT) {
+            // We need to update the header in case that the flipped bit was in there
+            memcpy(header, partition->device->_subpage_buffer, sizeof(subpageheader_s));
+        }
+
+        data_buffer += ecc_size;
+    }
+
+    data_buffer += sizeof(subpageheader_s);
+
+    if(buffer != 0) {
+        memcpy(buffer, data_buffer, header->data_length);
+    }
+
+    return 0;
+}
+
+
+
+/* ================
+ * Metadata storage
+ * ================ */
+
+int ftl_write_metadata(ftl_device_s *device, const void *metadata, uint16_t length) {
+    if(sizeof(ftl_metadata_header_s) +
+       device->partition_count * sizeof(ftl_partition_s) +
+       length > ftl_data_per_subpage(device, true)) {
+
+        return -EFBIG;
+    }
+
+    ftl_metadata_header_s header = {
+        .version = device->metadata_version+1,
+        .foreign_metadata_length = length,
+        .partition_count = device->partition_count
+    };
+
+    //printf("version: %x length %x count %x\n", header.version, header.foreign_metadata_length, header.partition_count);
+
+    int offset = sizeof(subpageheader_s) + device->ecc_size;
+    memset(device->_subpage_buffer, 0, device->subpage_size);
+
+    memcpy(device->_subpage_buffer + offset, &header, sizeof(ftl_metadata_header_s));
+    offset += sizeof(ftl_metadata_header_s);
+
+    for(unsigned int i=0; i<device->partition_count; i++) {
+        memcpy(device->_subpage_buffer + offset, device->partitions[i], sizeof(ftl_partition_s));
+        offset += sizeof(ftl_partition_s);
+    }
+
+    memcpy(device->_subpage_buffer + offset, metadata, length);
+    offset += length;
+
+    ftl_partition_s *metadata_partition = device->partitions[0];
+    _prepare_subpage_buffer_with_ecc(metadata_partition, offset+1);
+
+    //HEXDUMP_BUFFER(device->_subpage_buffer, device->subpage_size);
+
+    int ret = ftl_write_raw(metadata_partition, device->_subpage_buffer, metadata_partition->next_subpage);
+    metadata_partition->next_subpage++;
+    device->metadata_version++;
+    return ret;
+}
+
+// TODO: use binary search based mechanism
+int ftl_load_metadata_page_with_version(ftl_device_s *device, uint32_t version, uint32_t *source_page) {
+    ftl_partition_s *metadata_partition = device->partitions[0];
+
+    ftl_metadata_header_s metadata_header;
+    subpageheader_s subpage_header;
+    int ret;
+    for(uint32_t i=0; i<ftl_subpages_in_partition(metadata_partition); i++) {
+        ret = ftl_read(metadata_partition, NULL, &subpage_header, i);
+
+        //printf("reading page %d", i);
+        //HEXDUMP_BUFFER(device->_subpage_buffer, device->subpage_size);
+
+        // first page was not initialized, assuming empty partition
+        if(ret == -ENOENT && i==0) {
+            return -1;
+        }
+
+        //printf("xxx subpage %d error %d\n", i, ret);
+        if(ret != 0) {
```

Comment:
should be `if (ret > 0) {` or not? Because `ftl_read` may also return `EFAULT`.

Btw. coding convention suggests to use spaces before parentheses.
[edit] at least for `if`, `while` statements and the like, not everywhere[/edit]